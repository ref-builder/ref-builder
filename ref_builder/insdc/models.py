"""Models for data following INSDC standrds."""

import contextlib
from collections.abc import Iterator
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    RootModel,
    field_validator,
)

from ref_builder.models import Molecule, MolType


class FTGenbankLexicon(StrEnum):
    """Table keys used in Genbank."""

    FEATURE_KEY = "GBFeature_key"
    FEATURE_INTERVALS = "GBFeature_intervals"
    FEATURE_QUALIFIERS = "GBFeature_quals"
    INTERVAL_FROM = "GBInterval_from"
    INTERVAL_TO = "GBInterval_to"
    QUALIFIER_NAME = "GBQualifier_name"
    QUALIFIER_VALUE = "GBQualifier_value"


class INSDCDatabaseCrossReference(BaseModel):
    """A cross reference qualifier.

    Based on the INSDC guidelines for the /db_xref qualifier

    Reference:
    https://www.insdc.org/submitting-standards/dbxref-qualifier-vocabulary/
    """

    database: str

    identifier: int | str

    @classmethod
    def from_string(cls, db_xref: str) -> "INSDCDatabaseCrossReference":
        """Return labelled data from a standard ``db_xref`` string."""
        database, identifier = db_xref.split(":")[:2]

        return INSDCDatabaseCrossReference.model_validate(
            {"database": database, "identifier": identifier}
        )

    @field_validator("identifier", mode="after")
    @classmethod
    def convert_integer_identifier(cls, v: str | int) -> str | int:
        """Check if identifier can be converted to an integer and convert if so."""
        with contextlib.suppress(ValueError):
            return int(v)

        return v


class INSDCMolType(StrEnum):
    """In vivo molecule types.

    Based on the INSDC controlled vocabulary list for the /mol_type qualifier

    Reference:
    https://www.insdc.org/submitting-standards/controlled-vocabulary-moltype-qualifier/
    """

    GENOMIC_DNA = "genomic DNA"
    OTHER_DNA = "other DNA"
    UNASSIGNED_DNA = "unassigned DNA"
    GENOMIC_RNA = "genomic RNA"
    MRNA = "mRNA"
    TRNA = "tRNA"
    TRANSCRIBED_RNA = "transcribed RNA"
    VIRAL_CRNA = "viral cRNA"
    OTHER_RNA = "other RNA"

    @classmethod
    def from_molecule(cls: type, molecule: Molecule) -> "INSDCMolType":
        """Return an INSDC moltype value that matches the passed ref-builder molecule."""

        match molecule.type:
            case MolType.DNA:
                return INSDCMolType.GENOMIC_DNA
            case MolType.RNA:
                return INSDCMolType.GENOMIC_RNA
            case MolType.CRNA:
                return INSDCMolType.VIRAL_CRNA
            case MolType.MRNA:
                return INSDCMolType.MRNA
            case MolType.TRNA:
                return INSDCMolType.TRANSCRIBED_RNA
            case _:
                raise ValueError(
                    f"Molecule cannot be matched to INSDCMolType: {molecule.type}"
                )

    def to_molecule(self) -> MolType:
        """Return ref-builder molecule type."""
        if "DNA" in self.value:
            return MolType.DNA

        match self:
            case INSDCMolType.GENOMIC_RNA:
                return MolType.RNA
            case INSDCMolType.VIRAL_CRNA:
                return MolType.CRNA
            case INSDCMolType.MRNA:
                return MolType.MRNA
            case INSDCMolType.TRANSCRIBED_RNA:
                return MolType.TRNA
            case _:
                raise ValueError(
                    f"INSDC molecule type cannot be matched to MolType: {self}"
                )


class INSDCFeatureQualifier(BaseModel):
    """A feature qualifier in a GenBank source table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    name: Annotated[str, Field(validation_alias=FTGenbankLexicon.QUALIFIER_NAME)]
    """The name of the qualifier."""

    value: Annotated[
        str | None, Field(validation_alias=FTGenbankLexicon.QUALIFIER_VALUE)
    ] = None
    """The value of the qualifier. 
    If this field is not present, the qualifier is a flag.
    """


class INSDCFeatureInterval(BaseModel):
    """An interval in a GenBank source table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    start: Annotated[
        int,
        Field(validation_alias=FTGenbankLexicon.INTERVAL_FROM),
    ]
    """The start of the interval."""

    end: Annotated[
        int,
        Field(validation_alias=FTGenbankLexicon.INTERVAL_TO),
    ]
    """The end of the interval."""


class INSDCFeature(BaseModel):
    """An item in a feature table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    key: Annotated[
        str,
        Field(validation_alias=FTGenbankLexicon.FEATURE_KEY),
    ]
    """The name of the feature."""

    intervals: Annotated[
        list[INSDCFeatureInterval],
        Field(validation_alias=FTGenbankLexicon.FEATURE_INTERVALS),
    ]
    """The intervals of the feature."""

    qualifiers: Annotated[
        list[INSDCFeatureQualifier] | None,
        Field(validation_alias=FTGenbankLexicon.FEATURE_QUALIFIERS),
    ] = None
    """The qualifiers of the feature."""


class INSDCFeatureSource(INSDCFeature):
    """Normalized data from a source feature."""

    key: Annotated[
        Literal["source"],
        Field(validation_alias=FTGenbankLexicon.FEATURE_KEY),
    ]

    qualifiers: Annotated[
        list[INSDCFeatureQualifier],
        Field(validation_alias=FTGenbankLexicon.FEATURE_QUALIFIERS),
    ]


class INSDCFeatureTable(RootModel):
    """An unsorted INSDC feature table."""

    root: list[INSDCFeature]

    def __iter__(self) -> Iterator[INSDCFeature]:
        return iter(self.root)
