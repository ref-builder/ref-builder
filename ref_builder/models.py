from dataclasses import dataclass
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict


@dataclass
class OTUMinimal:
    """A minimal representation of an OTU."""

    id: UUID
    acronym: str
    legacy_id: str | None
    name: str
    taxid: int


class MolType(StrEnum):
    """The in vivo molecule type of a sequence.

    Corresponds to Genbank's moltype field
    """

    CRNA = "cRNA"
    DNA = "DNA"
    MRNA = "mRNA"
    RNA = "RNA"
    TRNA = "tRNA"


class Strandedness(StrEnum):
    """Strandedness of a molecule, either single or double."""

    SINGLE = "single"
    DOUBLE = "double"


class Topology(StrEnum):
    """Topology of a molecule, either linear or circular."""

    LINEAR = "linear"
    CIRCULAR = "circular"


class Molecule(BaseModel):
    """The strandedness, molecule type, and topology of an OTU."""

    model_config = ConfigDict(use_enum_values=True)

    strandedness: Strandedness
    topology: Topology
    type: MolType
