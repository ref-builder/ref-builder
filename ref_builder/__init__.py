"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
