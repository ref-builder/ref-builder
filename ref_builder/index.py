"""An index for improving repository performance."""

import binascii
import datetime
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from uuid import UUID

from ref_builder.events.base import EventMetadata
from ref_builder.models.otu import OTUMinimal
from ref_builder.otu.builders.otu import OTUBuilder


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: OTUBuilder
    """The OTU that was snapshotted."""


def adapt_uuid(original_uuid: UUID) -> str:
    """Automatically convert UUID to SQLite-compatible string."""
    return str(original_uuid)


def adapt_datetime(original_datetime: datetime.datetime) -> str:
    """Automatically convert datetime to SQLite-compatible string."""
    return original_datetime.isoformat()


def convert_uuid(string_uuid: bytes) -> UUID:
    """Automatically convert stringified UUID back to UUID type."""
    return UUID(string_uuid.decode())


def convert_datetime(string_datetime: bytes) -> datetime.datetime:
    """Automatically convert stringified datetime back to datetime type."""
    return datetime.datetime.fromisoformat(string_datetime.decode())


sqlite3.register_adapter(UUID, adapt_uuid)
sqlite3.register_adapter(datetime.datetime, adapt_datetime)

sqlite3.register_converter("uuid", convert_uuid)
sqlite3.register_converter("datetime", convert_datetime)


class Index:
    """An index for rapid access to repository data."""

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(
            path,
            isolation_level=None,
            detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES,
        )

        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT,
                timestamp TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS isolates (
                id TEXT PRIMARY KEY,
                name TEXT,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otu_updates (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                otu_id TEXT,
                timestamp_complete TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("isolates", "id"),
            ("otus", "id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
            ("otu_updates", "otu_id"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {
            row[0] for row in self.con.execute('SELECT id AS "id [uuid]" FROM otus')
        }

    def add_event_id(
        self,
        event_id: int,
        otu_id: UUID,
        timestamp: datetime.datetime,
    ) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        :param timestamp: the timestamp of the event
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id, timestamp)
            VALUES (?, ?, ?)
            """,
            (
                event_id,
                otu_id,
                timestamp,
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids,
                otu_id,
            )

        return None

    def get_first_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the earliest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_latest_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the latest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_acronym(self, acronym: str) -> UUID | None:
        """Get an OTU ID by its acronym."""
        if not acronym:
            raise ValueError("Acronym must contain at least one character.")

        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE acronym = ?',
            (acronym,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE name = ?',
            (name,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE taxid = ?',
            (taxid,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_isolate_id(self, isolate_id: UUID) -> UUID | None:
        """Get an OTU ID from an isolate ID that belongs to it."""
        cursor = self.con.execute(
            'SELECT otu_id AS "otu_id [uuid]" FROM isolates WHERE id = ?',
            (isolate_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM isolates WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.commit()

    def iter_event_metadata(self) -> Iterator[EventMetadata]:
        """Iterate over event metadata."""
        rows = self.con.execute(
            "SELECT event_id, otu_id, timestamp FROM events ORDER BY event_id",
        ).fetchall()

        for row in rows:
            yield EventMetadata(
                id=row[0],
                otu_id=UUID(row[1]) if row[1] else None,
                timestamp=row[2],
            )

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        rows = self.con.execute(
            """
            SELECT acronym, id AS "id [uuid]", name, taxid
            FROM otus ORDER BY name
            """,
        ).fetchall()

        for row in rows:
            yield OTUMinimal(
                acronym=row[0],
                id=row[1],
                name=row[2],
                taxid=row[3],
            )

    def get_last_otu_update_timestamp(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the last event in an update for an OTU."""
        cursor = self.con.execute(
            """
            SELECT timestamp_complete FROM otu_updates
            WHERE otu_id = ? ORDER BY id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return datetime.datetime.fromisoformat(result[0])

        return None

    def add_otu_update_history_entry(
        self, otu_id: UUID, timestamp: datetime.datetime
    ) -> int | None:
        """Write an entry into the OTU update history."""
        self.con.execute(
            """
            INSERT INTO
            otu_updates(otu_id, timestamp_complete)
            VALUES(?, ?)
            """,
            (
                otu_id,
                timestamp,
            ),
        )

        cursor = self.con.execute(
            "SELECT id FROM otu_updates WHERE otu_id = ? ORDER BY id DESC",
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (otu_id,),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = OTUBuilder.model_validate_json(otu_json)

        sequence_ids = [
            sequence.id for isolate in otu.isolates for sequence in isolate.sequences
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu.isolates:
            for sequence in isolate.sequences:
                sequence_id = str(sequence.id)

                if sequence_id in sequence_map:
                    sequence.sequence = sequence_map[sequence_id]
                else:
                    missing_sequences.append(str(sequence_id))

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=OTUBuilder.model_validate(otu),
        )

    def upsert_otu(self, otu: OTUBuilder, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.
        """
        sequence_ids = {seq.id for isolate in otu.isolates for seq in isolate.sequences}

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({placeholders});
            """,
            (otu.id, *sequence_ids),
        )

        batch = []

        # Insert or update the isolates.
        for isolate in otu.isolates:
            self.con.execute(
                "INSERT OR REPLACE INTO isolates (id, name, otu_id) VALUES (?, ?, ?)",
                (
                    isolate.id,
                    str(isolate.name),
                    otu.id,
                ),
            )

            for sequence in isolate.sequences:
                crc = _calculate_crc32(sequence.sequence)
                batch.append(
                    (
                        sequence.id,
                        crc,
                        otu.id,
                        sequence.sequence,
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, name, otu, taxid)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                otu.id,
                otu.acronym,
                at_event,
                otu.name,
                otu.model_dump_json(),
                otu.taxid,
            ),
        )

        self.con.commit()

    def prune(self, event_id: int) -> None:
        """Rollback the index to a previous event.

        * Deletes all events after the given event ID.
        * Deletes all OTU snapshots that have ``at_event`` later than ``event_id``.

        Removed snapshots will be automatically updated next time an OTU is fetched from
        the repository.

        :param event_id: the event ID to rollback to
        """
        self.con.execute(
            """
            DELETE FROM events WHERE event_id > ?
            """,
            (event_id,),
        )

        self.con.execute(
            """
            DELETE FROM otus WHERE at_event > ?
            """,
            (event_id,),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)
