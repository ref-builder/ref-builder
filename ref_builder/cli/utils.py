import sys
import warnings
from uuid import UUID

import click

from ref_builder.errors import InvalidInputError, PartialIDConflictError
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.repo import Repo
from ref_builder.utils import OTUDeletedWarning

pass_repo = click.make_pass_decorator(Repo)

UUID_STRING_LENGTH = 36


def get_otu_from_identifier(repo: Repo, identifier: str) -> OTUBuilder:
    """Return an OTU id from the repo if identifier matches a single OTU.

    The identifier can either be a stringified UUIDv4, a truncated portion
    of a UUID, a NCBI Taxonomy ID or an acronym associated with the OTU.

    Raise a PartialIDConflictError if >1 OTU is found.

    :param repo: the repository to be searched.
    :param identifier: a non-UUID identifier.
        Can be an integer Taxonomy ID, acronym or truncated partial UUID.
    :return: the UUID of the OTU or ``None``
    """
    otu_id = None

    if len(identifier) == UUID_STRING_LENGTH:
        try:
            otu_id = UUID(identifier)
        except ValueError:
            otu_id = None

    elif identifier.isnumeric():
        try:
            taxid = int(identifier)
        except ValueError:
            pass
        else:
            otu_id = repo.get_otu_id_by_taxid(taxid)

    else:
        if (otu_id := repo.get_otu_id_by_acronym(identifier)) is not None:
            return repo.get_otu(otu_id)

        try:
            otu_id = repo.get_otu_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError as e:
            click.echo(
                e,
                err=True,
            )

    if otu_id is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    with warnings.catch_warnings(
        category=OTUDeletedWarning, record=True
    ) as warning_list:
        otu_ = repo.get_otu(otu_id)

    if otu_ is None:
        if warning_list:
            click.echo("OTU has been deleted.", err=True)

        else:
            click.echo("OTU not found.", err=True)

        sys.exit(1)

    return otu_


def get_otu_isolate_ids_from_identifier(repo: Repo, identifier: str) -> (UUID, UUID):
    """Return an isolate ID from the repo if identifier matches a single isolate.

    Handles cases where the isolate cannot be found.
    """
    isolate_id = None

    if len(identifier) == 32:
        try:
            isolate_id = UUID(identifier)
        except ValueError:
            isolate_id = None

    else:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError:
            click.echo(
                "Partial ID segment must be at least 8 characters long.", err=True
            )
            sys.exit(1)

    if isolate_id is not None:
        if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is not None:
            if isolate_id in repo.get_otu(otu_id).isolate_ids:
                return otu_id, isolate_id

    click.echo("Isolate ID could not be found.", err=True)
    sys.exit(1)
