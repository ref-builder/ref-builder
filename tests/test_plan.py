from uuid import UUID, uuid4

import pytest
from pydantic import ValidationError
from syrupy import SnapshotAssertion

from ref_builder.ncbi.models import NCBISourceMolType
from ref_builder.plan import (
    Plan,
    Segment,
    SegmentName,
    SegmentRule,
    extract_segment_name_from_record,
    parse_segment_name,
)
from tests.fixtures.factories import NCBIGenbankFactory, NCBISourceFactory
from tests.fixtures.utils import uuid_matcher


class TestPlan:
    example: dict

    @pytest.fixture(autouse=True)
    def _setup(self):
        self.example = {
            "id": uuid4(),
            "segments": [
                {
                    "id": uuid4(),
                    "name": None,
                    "rule": "required",
                    "length": 100,
                    "length_tolerance": 0.01,
                }
            ],
        }

    def test_ok(self, snapshot: SnapshotAssertion):
        """Test that a plan can be created."""
        segment = self.example["segments"][0]

        plan = Plan.model_validate(self.example)

        assert (
            Plan(
                id=self.example["id"],
                segments=[
                    Segment(
                        id=segment["id"],
                        name=segment["name"],
                        rule=SegmentRule(segment["required"]),
                        length=segment["length"],
                        length_tolerance=segment["length_tolerance"],
                    )
                ],
            ).model_dump()
            == plan.model_dump()
            == snapshot(matcher=uuid_matcher)
        )

        assert plan.monopartite

    def test_new(self, snapshot: SnapshotAssertion):
        """Test that the new method creates a new plan without a UUID being provided."""
        plan = Plan.new(
            [
                Segment.new(100, 0.01, SegmentName(prefix="DNA", key="A")),
                Segment.new(200, 0.02, SegmentName(prefix="DNA", key="B")),
            ]
        )

        assert isinstance(plan.id, UUID)

        assert plan.model_dump() == snapshot(
            matcher=uuid_matcher,
        )

    @pytest.mark.parametrize("required", ["required", "recommended", "optional"])
    def test_segment_rule(self, required: str):
        """Test that a string segment rule can be parsed into  SegmentRule."""
        plan = Plan.model_validate(
            {
                **self.example,
                "segments": [
                    {
                        **self.example["segments"][0],
                        "required": required,
                    }
                ],
            }
        )

        assert plan.segments[0].rule == SegmentRule(required)

    @pytest.mark.parametrize(
        "name",
        [
            None,
            {"prefix": "DNA", "key": "A                                             "},
        ],
        ids=["two_unnamed", "one_unnamed"],
    )
    def test_multipartite_unnamed(self, name: dict | None):
        """Test that a multipartite path with any unnamed segments fails validation."""
        self.example["segments"].append({**self.example["segments"][0], "name": name})

        with pytest.raises(
            ValidationError,
            match="All segments must have a name in a multipartite plan.",
        ):
            Plan.model_validate(self.example)

    def test_duplicate_segment_names(self):
        """Test that a plan with duplicate segment names fails validation."""
        self.example["segments"] = [
            {**self.example["segments"][0], "name": {"prefix": "DNA", "key": "A"}}
            for _ in range(2)
        ]

        with pytest.raises(
            ValidationError,
            match="Segment names must be unique within a plan.",
        ):
            Plan.model_validate(self.example)

    def test_get_segment_by_id(self):
        """Test that a segment can be retrieved by its ID."""
        segment = self.example["segments"][0]

        plan = Plan.model_validate(self.example)

        assert (
            plan.get_segment_by_id(segment["id"]).model_dump()
            == Segment.model_validate(segment).model_dump()
        )

    def test_get_segment_by_nonexistent_id(self):
        """Test that None is returned when a segment is not found by its ID."""
        plan = Plan.model_validate(self.example)

        assert plan.get_segment_by_id(uuid4()) is None

    def test_monopartite(self):
        """Test that the monopartite property works as expected."""
        assert Plan.model_validate(self.example).monopartite is True

        self.example["segments"] = [
            {**self.example["segments"][0], "name": {"prefix": "DNA", "key": "A"}}
        ]

        assert Plan.model_validate(self.example).monopartite is True

        self.example["segments"].append(
            {**self.example["segments"][0], "name": {"prefix": "DNA", "key": "B"}}
        )

        assert Plan.model_validate(self.example).monopartite is False

    @pytest.mark.parametrize("rule", ["required", "recommended", "optional"])
    def test_required_segments_monopartite(self, rule: str):
        """Test that the required_segments property works as expected when a monopartite
        plan has either one or no required segments.
        """
        self.example["segments"][0]["required"] = rule

        plan = Plan.model_validate(self.example)

        if rule == "required":
            assert plan.not_required_segments == []
            assert plan.required_segments == plan.segments

        else:
            assert plan.not_required_segments == plan.segments
            assert plan.required_segments == []

    def test_required_segments_multipartite(self):
        """Test that the required_segments property works as expected."""
        self.example["segments"] = [
            {
                **self.example["segments"][0],
                "required": rule,
                "name": {"prefix": "DNA", "key": key},
            }
            for key, rule in zip(
                "ABCD", ["required", "recommended", "required", "optional"], strict=True
            )
        ]

        plan = Plan.model_validate(self.example)

        assert plan.not_required_segments == [plan.segments[1], plan.segments[3]]
        assert plan.required_segments == [plan.segments[0], plan.segments[2]]


class TestParseSegmentName:
    """Test segment name normalization."""

    @pytest.mark.parametrize("delimiter", [" ", "_", "-"])
    @pytest.mark.parametrize("key", ["A", "BN", "U3"])
    @pytest.mark.parametrize("prefix", ["DNA", "RNA"])
    def test_ok(self, delimiter: str, key: str, prefix: str):
        """Test that parse_segment_name() correctly parses prefix-key segment names."""
        assert parse_segment_name(f"{prefix}{delimiter}{key}") == SegmentName(
            prefix=prefix, key=key
        )

    @pytest.mark.parametrize(
        "string",
        ["", "*V/", "51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d"],
    )
    def test_none(self, string: str) -> None:
        """Test that parse_segment_name() returns expected ValueError for bad input."""
        assert parse_segment_name(string) is None


class TestExtractSegmentNameFromRecord:
    @pytest.mark.parametrize("key", ["1", "A", "BN", "U3"])
    @pytest.mark.parametrize(
        "mol_type", [NCBISourceMolType.GENOMIC_RNA, NCBISourceMolType.GENOMIC_DNA]
    )
    def test_ok(
        self,
        key: str,
        mol_type: NCBISourceMolType,
        ncbi_genbank_factory: NCBIGenbankFactory,
        ncbi_source_factory: NCBISourceFactory,
    ) -> None:
        """Test that a full segment name can be parsed
        when the source table segment field contains a key with no prefix.
        """
        record = ncbi_genbank_factory.build(
            source=ncbi_source_factory.build(
                mol_type=mol_type,
                segment=key,
            ),
        )

        match mol_type:
            case NCBISourceMolType.GENOMIC_DNA:
                suffix = "DNA"
            case NCBISourceMolType.GENOMIC_RNA:
                suffix = "RNA"
            case _:
                raise ValueError(f"{mol_type} may not be a valid NCBISourceMolType.")

        assert extract_segment_name_from_record(record) == SegmentName(
            prefix=suffix, key=key
        )

    @pytest.mark.parametrize("key", ["", "V#", "51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d"])
    @pytest.mark.parametrize(
        "mol_type", [NCBISourceMolType.GENOMIC_RNA, NCBISourceMolType.GENOMIC_DNA]
    )
    def test_parse_from_record_fail(
        self,
        key: str,
        mol_type: NCBISourceMolType,
        ncbi_genbank_factory: NCBIGenbankFactory,
        ncbi_source_factory: NCBISourceFactory,
    ):
        """Test that ``None`` is returned if no segment name is found."""
        record = ncbi_genbank_factory.build(
            source=ncbi_source_factory.build(
                mol_type=mol_type,
                segment=key,
            ),
        )

        assert extract_segment_name_from_record(record) is None
