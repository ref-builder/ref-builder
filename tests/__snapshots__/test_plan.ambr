# serializer version: 1
# name: TestPlan.test_new
  dict({
    'id': UUID,
    'segments': list([
      dict({
        'id': UUID,
        'length': 100,
        'length_tolerance': 0.01,
        'name': dict({
          'key': 'A',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 200,
        'length_tolerance': 0.02,
        'name': dict({
          'key': 'B',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
    ]),
  })
# ---
# name: TestPlan.test_ok
  dict({
    'id': UUID,
    'segments': list([
      dict({
        'id': UUID,
        'length': 100,
        'length_tolerance': 0.01,
        'name': None,
        'rule': 'required',
      }),
    ]),
  })
# ---
